### PR TITLE
[Dev] feat: m4 leftover changes

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do ?

Main branch pr: https://github.com/NVIDIA/Megatron-LM/pull/2506

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/distributed/distributed_data_parallel.py [hunk 1]
```diff
@@ -55,10 +55,15 @@ def __init__(
         # If using very large dp_sizes, make buckets larger to ensure that chunks used in NCCL
         # ring-reduce implementations are large enough to remain bandwidth-bound rather than
         # latency-bound.
+        # Setup process groups using DDP-specific helper method (handles both None and provided pg_collection)
+        process_groups = ProcessGroupCollection.setup_process_groups_for_ddp(
```
Comment (skyw) [inline megatron/core/distributed/distributed_data_parallel.py]: Suggestion: Make naming more explicit, e.g. process_group_dict. Using plural form to differentiate with process_group can lead to very hard to spot typos.

File: megatron/core/extensions/transformer_engine.py [hunk 2]
```diff
@@ -453,7 +450,7 @@ def sharded_state_dict(self, prefix="", sharded_offsets=(), metadata=None):
             prefix,
             None,
             sharded_offsets,
-            tp_group=self._tp_group,
+            tp_group=None,
```
Comment (skyw) [inline megatron/core/extensions/transformer_engine.py]: Q: This seems a big change. Why it is needed before and not anymore?

File: megatron/core/optimizer/__init__.py [hunk 3]
```diff
@@ -577,10 +577,20 @@ def get_megatron_optimizer(
             )
             model_chunk_offset += 1
 
-        if len(optimizers) == 1:
-            return optimizers[0]
+        # Build final optimizer (single or chained)
+        result_optimizer = optimizers[0] if len(optimizers) == 1 else ChainedOptimizer(optimizers)
 
-        return ChainedOptimizer(optimizers)
+        tp_group = (
```
Comment (skyw) [inline megatron/core/optimizer/__init__.py]: Better to add comment of what the following code does and why it is needed.

File: megatron/core/optimizer/__init__.py [hunk 4]
```diff
@@ -671,4 +681,14 @@ def get_megatron_optimizer(
             state_dict=param_to_param_group, checkpoint_id=dump_param_to_param_group_map
         )
 
-    return ChainedOptimizer(optimizers)
+    # Build final optimizer (always chained in this branch, possibly with 1 element)
+    result_optimizer = ChainedOptimizer(optimizers)
+    # Attach TP group so duplicate-checks don't rely on legacy globals.
+    tp_group = (
```
Comment (skyw) [inline megatron/core/optimizer/__init__.py]: DRY. Code is not exactly the same but very close to the above one. any chance to simplify?

File: megatron/core/pipeline_parallel/schedules.py [hunk 5]
```diff
@@ -123,10 +123,18 @@ def forward_step(data_iterator, model):
...
+    Args:
+        pp_size (Optional[int]): Pipeline model parallel size to use.
+        vp_size (Optional[int]): Virtual pipeline model parallel size to use.
+            If both pp_size and vp_size are None, both values fall back to parallel_state.
+            Otherwise, provided values are used as-is and None is treated as an explicit input.
     """
-    pipeline_model_parallel_size = parallel_state.get_pipeline_model_parallel_world_size()
-    if pipeline_model_parallel_size > 1:
-        if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
+    if pp_size is None and vp_size is None:
```
Comment (skyw) [inline megatron/core/pipeline_parallel/schedules.py:133]: Add check to avoid one of them being passed.

File: megatron/core/transformer/moe/shared_experts.py [hunk 6]
```diff
@@ -43,7 +43,7 @@ def __init__(
         gate: bool,
         pg_collection: Optional[ProcessGroupCollection] = None,
     ):
-        config = deepcopy(config)
+        config = copy(config)
```
Comment (skyw) [inline megatron/core/transformer/moe/shared_experts.py:46]: This suggests deepcopy was a bug because behavior would be very different. Will need review from people are more familiar with this code.

Comment (yaoyu-33) [reply megatron/core/transformer/moe/shared_experts.py:46]: deepcopy on config containing pgs, funcs doesnt make too much sense. @yanring  can you find someone to check if deepcopy is actually required. I notice some other places are using deepcopy but feel totally unnecessary.

File: megatron/core/extensions/transformer_engine.py (hunk 2 above)
Comment (yaoyu-33) [reply megatron/core/extensions/transformer_engine.py]: oh my bad, needs to fix this. It's just a WAR to unblock moe, this is a complicated situation for moe saving.

File: megatron/core/optimizer/__init__.py (hunk 4 above)
Comment (yaoyu-33) [reply megatron/core/optimizer/__init__.py]: done

File: megatron/core/optimizer/__init__.py (hunk 3 above)
Comment (yaoyu-33) [reply megatron/core/optimizer/__init__.py]: done

File: megatron/core/distributed/distributed_data_parallel.py (hunk 1 above)
Comment (yaoyu-33) [reply megatron/core/distributed/distributed_data_parallel.py]: done

File: megatron/core/pipeline_parallel/schedules.py (hunk 5 above)
Comment (yaoyu-33) [reply megatron/core/pipeline_parallel/schedules.py:133]: vp_size can be None and is default to be None in some cases I think

**Review by skyw — APPROVED** [review]

File: megatron/core/transformer/moe/shared_experts.py (hunk 6 above)
Comment (yanring) [reply megatron/core/transformer/moe/shared_experts.py:46]: IIRC, this is because we modified the ffn_hidden_size in the config below. @hxbai

File: megatron/core/transformer/module.py [hunk 7]
```diff
@@ -438,11 +440,20 @@ def forward(self, *inputs, fp32_output=True, **kwargs):
...
         """
-        if parallel_state.is_pipeline_first_stage(ignore_virtual=False, vp_stage=self.vp_stage):
+        from megatron.core.pipeline_parallel.utils import (
+            is_pp_first_stage,
+            is_pp_last_stage,
+            is_vp_first_stage,
+            is_vp_last_stage,
+        )
+
+        pp_group = self.pg_collection.pp
```
Comment (yanring) [inline megatron/core/transformer/module.py]: self.pg_collection could be None

File: megatron/core/optimizer/__init__.py [hunk 8]
```diff
@@ -473,6 +473,34 @@ def init_state_fn(opt, config=None):
...
+    """Collapse the optimizer list and attach the tensor-parallel group.
+
+    Each Megatron optimizer instance needs the tensor-parallel group reference so
+    duplicate parameter safety checks can avoid legacy global state. This helper
+    makes that behavior explicit and keeps the construction path consistent.
+    """
+    if not optimizers:
+        raise RuntimeError("Cannot finalize optimizer chain with an empty optimizer list.")
+
+    final_optimizer = optimizers[0] if len(optimizers) == 1 else ChainedOptimizer(optimizers)
```
Comment (yanring) [inline megatron/core/optimizer/__init__.py]: This is kind of hacky, tp_group is not declared in the optimizer class definition, and object state is modified after construction.

I understand this is likely a pragmatic choice to minimize the diff size for M4 migration. If that's the case, could we add a # TODO(M4): comment in optimizer init noting that this should be refactored to pass tp_group through constructors in a follow-up PR?

**Comment by yanring** [comment]
LGTM, left a few comments. Will trigger merge after resolving them.

File: megatron/core/optimizer/__init__.py (hunk 8 above)
Comment (yaoyu-33) [reply megatron/core/optimizer/__init__.py]: agree I will add a todo for this.

File: megatron/core/transformer/moe/shared_experts.py (hunk 6 above)
Comment (yaoyu-33) [reply megatron/core/transformer/moe/shared_experts.py:46]: ffn_hidden_size should only requires copy from what I can see

File: megatron/core/transformer/module.py (hunk 7 above)
Comment (yaoyu-33) [reply megatron/core/transformer/module.py]: yeah, I will add a guard here.

File: megatron/core/optimizer/__init__.py (hunk 8 above)
Comment (yaoyu-33) [reply megatron/core/optimizer/__init__.py]: cleaned up a bit and moved todo

**Comment by ko3n1g** [comment]
/ok to test [7c1ed1a](https://github.com/NVIDIA/Megatron-LM/pull/2226/commits/7c1ed1a6a9fc65db257928f7f56b9a0c2acfadcc)
